### PR TITLE
Feature/alert/staking inactive

### PR DESCRIPTION
**Pull request by FrogTravel**

### CODE REVIEW

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/main/StakingViewModel.kt [hunk 1]
```diff
@@ -79,6 +90,16 @@ class StakingViewModel(
         }
     }
 
+    val alertsFlow = getAlertsFlow().mapList { AlertModel.mapAlertToAlertModel(it, router) }.asLiveData()
+
+    private fun getAlertsFlow(): Flow<List<Alert>> {
+        val stakingState = runBlocking {
```
Comment (FrogTravel) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/main/StakingViewModel.kt:96]: How do you think we can get rid of runBlocking in here? To get current stash acc I have to call coroutine function 
@valentunn

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        return if (context.isAllValidatorsNotElected) Alert.ChangeValidators else null
+    }
+
+    private val alertProducers = listOf(
+        ::produceElectionAlert,
+        ::produceValidatorsAlert
+    )
+
+    fun getAlertsFlow(stakingState: StakingState): Flow<List<Alert>> = flow {
+        val alertsFlow = combine(
```
Comment (valentunn) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt:37]: combine(
stakingRepository.electionFlow(networkType),
stakingRepository.electedExposuresInActiveEra,
) { electionStatus, exposures ->
val alertResolutionContext = AlertContext(electionStatus,  exposures )
}

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    private suspend fun getElectionStatus(networkType: Node.NetworkType): Flow<Boolean> {
+        return stakingRepository.electionFlow(networkType).map {
+            it != Election.OPEN
+        }
+    }
+
+    private suspend fun getChangeValidators(stakingState: StakingState.Stash) = flow {
+        val exposures = stakingRepository.electedExposuresInActiveEra.first().values
+        val allNominators = exposures.flatMap { it.others }.map { it.who }
+        val result = allNominators.find { it.contentEquals(stakingState.stashId) } != null
```
Comment (valentunn) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt:57]: exposures.any { exposure ->  exposure.others.any { it.who.contentEquals(stakingState.stashId)}}

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+class AlertModel(
+    @DrawableRes val icon: Int,
+    @StringRes val title: Int,
+    @StringRes val extraMessage: Int,
+    val startFlow: ((View) -> Unit)? = null,
+) {
+    val isWarning: Boolean = startFlow == null
+
+    companion object {
+        fun mapAlertToAlertModel(alert: Alert, router: StakingRouter): AlertModel {
```
Comment (valentunn) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt:19]: move to viewmodel

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import jp.co.soramitsu.feature_staking_impl.domain.Alert
+import jp.co.soramitsu.feature_staking_impl.presentation.StakingRouter
+
+class AlertModel(
+    @DrawableRes val icon: Int,
+    @StringRes val title: Int,
+    @StringRes val extraMessage: Int,
+    val startFlow: ((View) -> Unit)? = null,
+) {
+    val isWarning: Boolean = startFlow == null
```
Comment (valentunn) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt:16]: Make sealed structure.
Move all logic to mapper

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertStatus.kt [hunk 6]
```diff
@@ -0,0 +1,7 @@
+package jp.co.soramitsu.feature_staking_impl.presentation.staking.alerts.model
+
+sealed class AlertStatus {
+    class Alerts(val alerts: List<AlertModel>) : AlertStatus()
+
+    object NoAlerts : AlertStatus()
+}
```
Comment (valentunn) [inline feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertStatus.kt:7]: Just List<AlertModel>

File: feature-staking-impl/src/main/res/layout/view_alert.xml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            android:layout_height="0dp"
+            android:layout_marginStart="16dp"
+            android:layout_marginTop="8dp"
+            android:layout_weight="1"
+            android:text="@string/staking_alert_no_alerts_now" />
+
+        <androidx.recyclerview.widget.RecyclerView
+            android:id="@+id/alertsRecycler"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
```
Comment (valentunn) [inline feature-staking-impl/src/main/res/layout/view_alert.xml:49]: wrap_content

File: feature-staking-impl/src/main/res/layout/view_alert.xml [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+<merge xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    tools:orientation="vertical"
+    tools:background="@color/black"
+    tools:parentTag="android.widget.LinearLayout">
+
+    <LinearLayout
```
Comment (valentunn) [inline feature-staking-impl/src/main/res/layout/view_alert.xml:11]: Redundant inner container

**Review by valentunn — CHANGES_REQUESTED** [review]

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/main/StakingViewModel.kt (hunk 1 above)
Comment (valentunn) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/main/StakingViewModel.kt:96]: flatMapLatest on selectedAccountStakingStateFlow

File: common/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -1,5 +1,14 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <resources>
+    <string name="staking_alert_title">Alert</string>
+    <string name="staking_alert_no_alerts_now">Everything is fine now. Alerts will appear here.</string>
+    <string name="staking_alert_start_next_era">Your staking will start in the next era.</string>
+    <string name="staking_alert_start_next_era_extra_message">Please wait for the next era to start.</string>
+    <string name="staking_alert_election">Staking actions are currently unavailable.</string>
+    <string name="staking_alert_start_election_extra_message">The network is electing validators. Usually it takes less than 10 minutes.</string>
+    <string name="staking_alert_change_validators">Change your validators.</string>
+    <string name="staking_alert_change_validators_extra_message">Staking was inactive. None of yourvalidators were elected by network.</string>
```
[the remark this answers is not in the document]
Comment (FrogTravel) [reply common/src/main/res/values/strings.xml]: Done

File: feature-staking-api/src/main/java/jp/co/soramitsu/feature_staking_api/domain/api/StakingRepository.kt [hunk 10]
```diff
@@ -47,6 +48,8 @@ interface StakingRepository {
     suspend fun getRewardDestination(stakingState: StakingState.Stash): RewardDestination
 
     suspend fun getControllerAccountInfo(stakingState: StakingState.Stash): AccountInfo
+
+    val electedExposuresInActiveEra: SharedFlow<Map<String, Exposure>>
```
[the remark this answers is not in the document]
Comment (FrogTravel) [reply feature-staking-api/src/main/java/jp/co/soramitsu/feature_staking_api/domain/api/StakingRepository.kt:52]: Done

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt (hunk 2 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt:42]: Done

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt (hunk 3 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/domain/AlertsInteractor.kt]: Done

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt (hunk 4 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt]: Done

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt (hunk 5 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertModel.kt]: Done

File: feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertStatus.kt (hunk 6 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/java/jp/co/soramitsu/feature_staking_impl/presentation/staking/alerts/model/AlertStatus.kt]: Done

File: feature-staking-impl/src/main/res/layout/view_alert.xml (hunk 7 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/res/layout/view_alert.xml]: Done

File: feature-staking-impl/src/main/res/layout/view_alert.xml (hunk 8 above)
Comment (FrogTravel) [reply feature-staking-impl/src/main/res/layout/view_alert.xml]: Done

**Review by valentunn — APPROVED** [review]

**Pull request merged**